### PR TITLE
Add Copilot runtime connection support and update documentation

### PR DESCRIPTION
**Pull request by Shazwazza**
This pull request adds support for connecting the Copilot provider to an already-running Copilot runtime, rather than always spawning a nested process. This enables external orchestrators to share a single authenticated Copilot session across all agents and models, with configuration available via YAML or environment variables. The schema enforces mutual exclusivity between this connection method and custom endpoint routing, and provides clear error handling and documentation.

**Copilot existing runtime connection support:**

* `ProviderSettings` schema (`src/conductor/config/schema.py`): Added `runtime_url` and `runtime_token` fields to allow connecting to an already-running Copilot runtime instead of spawning a new one. These fields can be set via YAML or the environment (`COPILOT_PROVIDER_RUNTIME_URL`, `COPILOT_PROVIDER_RUNTIME_TOKEN`). [[1]](diffhunk://#diff-2f6500f999b60c85af0c4e10fd9ad460b1f1a48f26dddc6c58176d556e4860a8R1714-R1747) [[2]](diffhunk://#diff-2f6500f999b60c85af0c4e10fd9ad460b1f1a48f26dddc6c58176d556e4860a8R1793-R1794)
* Schema validation (`src/conductor/config/schema.py`): Enforces that `runtime_url` is mutually exclusive with custom endpoint routing fields, and that `runtime_token` requires `runtime_url`. Empty tokens are rejected. [[1]](diffhunk://#diff-2f6500f999b60c85af0c4e10fd9ad460b1f1a48f26dddc6c58176d556e4860a8R1845) [[2]](diffhunk://#diff-2f6500f999b60c85af0c4e10fd9ad460b1f1a48f26dddc6c58176d556e4860a8R1877-R1905)
* Added `has_external_runtime()` method to `ProviderSettings` to distinguish this mode from custom routing.

**Documentation and examples:**

* Documentation (`docs/configuration.md`, `AGENTS.md`): Added detailed documentation and step-by-step instructions for connecting to an existing Copilot runtime, including configuration via YAML or environment variables, rules, and usage notes. [[1]](diffhunk://#diff-17ed18489a956f326ec0fe4040850c5bc9261d4631fb42da4c52891d74a59180R236-R312) [[2]](diffhunk://#diff-a54ff182c7e8acf56acfd6e4b9c3ff41e2c41a31c9b211b2deb9df75d9a478f9L114-R114) [[3]](diffhunk://#diff-a54ff182c7e8acf56acfd6e4b9c3ff41e2c41a31c9b211b2deb9df75d9a478f9R157)
* Example workflow (`examples/copilot-existing-runtime.yaml`): Added a new example YAML workflow demonstrating how to connect to an already-running Copilot runtime.
* Changelog (`CHANGELOG.md`): Added entry describing the new feature and its configuration.

### CODE REVIEW

**Comment by Shazwazza** [comment]
# 🤖 Multi-Agent PR Review #1

**PR**: #283 — Add Copilot runtime connection support and update documentation
**Merge readiness**: ⚠️ Needs minor fixes
**Review confidence**: 71/100 🟡 Moderate · *reviewer agreement & finding strength — not a PR-quality score*
**Models**: claude-opus-4.8 (Security & Correctness), gpt-5.5 (Robustness & Completeness)
**Deliberation**: 1 round, consensus reached

---

## 📋 Summary

This PR adds a well-scoped, well-documented feature: connecting the Copilot provider to an already-running `copilot --server` runtime via `runtime_url` / `runtime_token` (YAML or namespaced env vars), with secrets handled as redacted `SecretStr` and mutual exclusivity with custom endpoint routing enforced at config load. Both reviewers agree the change is low-risk and production-safe once a handful of robustness gaps are tightened, and both landed on ⚠️ *Needs minor fixes*. All four findings reached consensus in a single round — there are **no disputed or unique findings**. The high-signal themes are (1) an SDK-import coupling that could silently disable the whole Copilot provider on an older-but-present SDK, and (2) the environment-variable activation path bypassing the same validation and mutual-exclusion guarantees the YAML path enforces. Consensus is genuine but weak on three of four items (Reviewer 1 held moderate conviction throughout), so treat these as sound-but-not-emphatic asks the author should still address.

## 🔍 Consensus Findings

All four findings below were agreed by both reviewers.

| Severity | File | Finding | Recommendation |
|----------|------|---------|----------------|
| 🟡 | `src/conductor/providers/copilot.py:82` | `from copilot.client import RuntimeConnection` sits inside the shared SDK-availability `try`. An installed SDK lacking that symbol (pin is only `>=1.0.0`) flips `COPILOT_SDK_AVAILABLE` to `False` and silently disables the entire Copilot provider — including the pre-existing nested-spawn path — for users who never opt into runtime connections. | Import `RuntimeConnection` in its own `try/except ImportError`; raise a clear non-retryable `ProviderError` only when a runtime connection is actually requested and it's unavailable. Bumping the pin to `>=1.0.1` is an acceptable add-on, not a substitute for graceful degradation. |
| 🟡 | `src/conductor/config/schema.py:1881-1904` | Empty `runtime_token`, `headers`, and `SecretStr` are rejected as likely typo/unset-env interpolation, but empty-string `runtime_url` is not. `""` is `not None`, so `has_external_runtime()` returns `True` and the `runtime_token requires runtime_url` guard passes for `runtime_url: ""` — then at runtime `""` is treated as falsy, falls back to env, and silently spawns a nested runtime while dropping the token. | Reject an empty `runtime_url` at schema validation, mirroring the empty-secret / empty-headers handling, so unset-env-interpolation typos fail loudly. |
| 🟡 | `src/conductor/providers/copilot.py:2082` | The schema rejects YAML `runtime_url` combined with custom endpoint routing, but the provider only logs a warning and proceeds when `runtime_url` arrives via `COPILOT_PROVIDER_RUNTIME_URL`. This bypasses the documented mutual-exclusion rule and can connect a custom-routed workflow to an external runtime from a polluted environment. The codebase is also internally inconsistent (schema comment says routing "would silently be ignored"; the warning says it "will still be applied"). | Enforce mutual exclusion after env fallback too: if an external runtime is resolved and `runtime.provider` has custom routing, raise a non-retryable `ProviderError` instead of warning. Land with a dedicated test. |
| 🟡 | `src/conductor/providers/copilot.py:493-497` | Env-resolved runtime values aren't validated like YAML values. An empty `COPILOT_PROVIDER_RUNTIME_URL` falls through to a silent nested spawn, and a token-only environment silently drops the token and spawns — contradicting the YAML path, which hard-rejects `runtime_token` without `runtime_url`. Both reviewers agree an empty `COPILOT_PROVIDER_RUNTIME_TOKEN` should normalize to `None` (legitimate no-auth runtime), not error. | Strip and validate env-resolved values before deciding to spawn: raise on empty-after-strip `runtime_url` and on token-without-url; normalize empty `runtime_token` to `None`. Add provider tests for empty-URL, token-only, and empty-token env combinations. |

<details>
<summary>💡 Suggested fix for R1-001</summary>

```diff
try:
    from copilot import CopilotClient
    from copilot.session import PermissionHandler

    COPILOT_SDK_AVAILABLE = True
except ImportError:
    COPILOT_SDK_AVAILABLE = False
    CopilotClient = None  # type: ignore[misc, assignment]
    PermissionHandler = None  # type: ignore[misc, assignment]

+# RuntimeConnection was added to the SDK after CopilotClient; import it
+# separately so an older-but-present SDK still enables the default provider.
+try:
+    from copilot.client import RuntimeConnection
+except ImportError:
+    RuntimeConnection = None  # type: ignore[misc, assignment]
```
</details>

<details>
<summary>💡 Suggested fix for R1-002</summary>

```diff
+        if self.runtime_url is not None:
+            if self.runtime_url == "":
+                raise ValueError(
+                    "'runtime_url' is empty; remove the key or supply a value "
+                    "(typo / unset env interpolation?)"
+                )
```
</details>

<details>
<summary>💡 Suggested fix for R2-001</summary>

```diff
        url, token = connection
+        if self._provider_settings is not None and self._provider_settings.has_custom_routing():
+            raise ProviderError(
+                "COPILOT_PROVIDER_RUNTIME_URL cannot be combined with custom endpoint "
+                "routing in runtime.provider; connecting to an existing runtime and "
+                "custom endpoint routing are mutually exclusive (the runtime is the endpoint).",
+                suggestion=(
+                    "Unset COPILOT_PROVIDER_RUNTIME_URL or remove base_url/api_key/"
+                    "bearer_token/type/wire_api/headers/azure from runtime.provider."
+                ),
+                is_retryable=False,
+            )
```
</details>

<details>
<summary>💡 Suggested fix for R2-002</summary>

```diff
        if token is None:
            token = os.environ.get("COPILOT_PROVIDER_RUNTIME_TOKEN")

+        if url is not None:
+            url = url.strip()
+            if not url:
+                raise ProviderError(
+                    "'runtime_url' is empty; remove it or supply a value.",
+                    suggestion=(
+                        "Set runtime.provider.runtime_url or COPILOT_PROVIDER_RUNTIME_URL "
+                        "to a valid port, host:port, or full URL."
+                    ),
+                    is_retryable=False,
+                )
+        if token is not None:
+            token = token.strip() or None
+        if token is not None and url is None:
+            raise ProviderError(
+                "'runtime_token' requires 'runtime_url' to also be set",
+                suggestion=(
+                    "Set COPILOT_PROVIDER_RUNTIME_URL alongside "
+                    "COPILOT_PROVIDER_RUNTIME_TOKEN, or remove the token."
+                ),
+                is_retryable=False,
+            )
+
        if url is None:
            return None
        return (url, token)
```
</details>

## 🔎 Unique / Disputed Findings

_None._ All four findings reached mutual consensus in round 1, with no open disagreements.

## 📊 Model Comparison

| Aspect | Reviewer 1 (claude-opus-4.8) | Reviewer 2 (gpt-5.5) |
|--------|-------------------|----------------------|
| Merge Readiness | ⚠️ Needs minor fixes | ⚠️ Needs minor fixes |
| Critical findings | 0 | 0 |
| Warnings | 2 | 2 |
| Suggestions | 0 | 0 |
| Total findings | 2 | 2 |

## 📐 Impact Diagram

The findings all center on the runtime-connection resolution decision (YAML → env fallback → spawn vs. connect, plus mutual-exclusion). This flow clarifies where each fix applies:

```mermaid
flowchart TD
    A[Provider start] --> B{runtime_url resolved?<br/>YAML then env}
    B -->|No| C[Spawn nested copilot runtime]
    B -->|Empty / typo| D[R1-002 &amp; R2-002:<br/>reject, don't silently spawn]
    B -->|Yes| E{Custom routing also active?}
    E -->|Yes via env| F[R2-001:<br/>raise ProviderError<br/>mutually exclusive]
    E -->|No| G[Connect to existing<br/>copilot --server runtime]
```

## 🏁 Verdict

Both reviewers agree: this is a clean, well-documented, low-risk feature that is production-safe once the robustness gaps are closed. There are no critical findings and no disputes — the four consensus warnings all cluster around SDK-import isolation and making the environment-variable activation path enforce the same validation and mutual-exclusion rules the YAML path already guarantees. None is a blocker, and each has a concrete suggested fix.

**Merge readiness**: ⚠️ Needs minor fixes — address the `RuntimeConnection` import isolation (R1-001) and the env-path validation/mutual-exclusion parity (R2-001, R2-002, R1-002) before merge.
**Review confidence**: 71/100 🟡 Moderate · *reviewer agreement & finding strength — not a PR-quality score*

File: src/conductor/providers/copilot.py [hunk 1]
```diff
@@ -79,12 +79,14 @@ def _is_finite_nonneg(value: object) -> TypeGuard[float]:
 # Try to import the Copilot SDK
 try:
     from copilot import CopilotClient
+    from copilot.client import RuntimeConnection
```
Comment (Shazwazza) [inline src/conductor/providers/copilot.py]: 🟡 **correctness**: [R1-001] The new `from copilot.client import RuntimeConnection` is placed inside the shared SDK-availability `try` block. The dependency is pinned only as `github-copilot-sdk>=1.0.0` (uv.lock resolves to 1.0.1). If an installed SDK version has `CopilotClient` and `PermissionHandler` but not `RuntimeConnection`, the whole `try` raises `ImportError`, `COPILOT_SDK_AVAILABLE` flips to `False`, and the entire Copilot provider — including the default nested-spawn path that predates this feature — is silently disabled for users who never opt into runtime connections.

**Recommendation**: Import `RuntimeConnection` in its own `try/except ImportError` block so an older-but-present SDK still enables the default provider. If a runtime connection is actually requested and `RuntimeConnection is None`, raise a clear non-retryable `ProviderError` suggesting an SDK upgrade; otherwise keep default `CopilotClient()` spawning enabled. Bumping the pin to `>=1.0.1` is acceptable as an additional cleanup but not a substitute for graceful degradation.

**Suggestion**:
```python
try:
    from copilot import CopilotClient
    from copilot.session import PermissionHandler

    COPILOT_SDK_AVAILABLE = True
except ImportError:
    COPILOT_SDK_AVAILABLE = False
    CopilotClient = None  # type: ignore[misc, assignment]
    PermissionHandler = None  # type: ignore[misc, assignment]

# RuntimeConnection was added to the SDK after CopilotClient; import it
# separately so an older-but-present SDK still enables the default provider.
try:
    from copilot.client import RuntimeConnection
except ImportError:
    RuntimeConnection = None  # type: ignore[misc, assignment]
```

**Review by Shazwazza — COMMENTED** [review]
🤖 Multi-Agent PR Review — Inline Findings

Posted by the pr-reviewer agent. See the summary comment for the full report.

File: src/conductor/config/schema.py [hunk 2]
```diff
@@ -1837,6 +1874,35 @@ def _check_field_compatibility(self) -> ProviderSettings:
                 "'azure' block is empty; either set azure.api_version or remove the block"
             )
 
+        # Connecting to an external runtime (runtime_url) is mutually
+        # exclusive with custom endpoint routing — the runtime *is* the
+        # endpoint, so a base_url / api_key / etc. alongside it is
+        # contradictory and would silently be ignored at the SDK boundary.
+        if self.runtime_url is not None:
```
Comment (Shazwazza) [inline src/conductor/config/schema.py]: 🟡 **correctness**: [R1-002] Empty `runtime_token`, empty `headers`, and empty `SecretStr` values are explicitly rejected as likely typo/unset-env-interpolation, but an empty-string `runtime_url` is not. Because `""` is `not None`, `has_external_runtime()` returns `True`, serialization stays structured, and mutual-exclusion runs — yet the `runtime_token requires runtime_url` guard (`self.runtime_url is None`) evaluates `False` for `""`, so `runtime_token: "secret"` + `runtime_url: ""` passes validation. At provider runtime `_resolve_runtime_connection()` treats `""` as falsy, falls back to env, and silently spawns a nested runtime while dropping the token.

**Recommendation**: Reject an empty `runtime_url` at schema validation, mirroring the empty-secret / empty-headers handling, so unset-env-interpolation typos fail loudly rather than silently falling through to spawn.

**Suggestion**:
```python
        if self.runtime_url is not None:
            if self.runtime_url == "":
                raise ValueError(
                    "'runtime_url' is empty; remove the key or supply a value "
                    "(typo / unset env interpolation?)"
                )
```

File: src/conductor/providers/copilot.py [hunk 3]
```diff
@@ -2027,6 +2064,35 @@ async def _ensure_client_started(self) -> None:
...
+        SDK's ``start()`` skips process spawning for URI connections and its
+        ``stop()`` leaves the externally-owned server running, so Conductor
+        reuses the server's single authenticated session for every agent.
+        """
+        connection = self._resolve_runtime_connection()
+        if connection is None:
+            return CopilotClient()
+
+        url, token = connection
+        if self._provider_settings is not None and self._provider_settings.has_custom_routing():
```
Comment (Shazwazza) [inline src/conductor/providers/copilot.py]: 🟡 **completeness**: [R2-001] The schema rejects YAML `runtime_url` combined with custom endpoint routing, but the provider layer only logs a warning and proceeds when `runtime_url` arrives via `COPILOT_PROVIDER_RUNTIME_URL` from the environment. This bypasses the documented mutual-exclusion rule and can unexpectedly connect a custom-routed workflow to an external runtime when the environment is polluted by an orchestrator or previous shell state. The codebase is also internally inconsistent about the effect (schema comment says custom routing 'would silently be ignored'; the warning says it 'will still be applied'), which is further reason to fail fast.

**Recommendation**: Enforce the mutual exclusion after environment fallback resolution as well. If an external runtime connection is resolved and `runtime.provider` has custom routing enabled, raise a non-retryable `ProviderError` instead of logging a warning and proceeding. Land this behavior-tightening with a dedicated test.

**Suggestion**:
```python
        url, token = connection
        if self._provider_settings is not None and self._provider_settings.has_custom_routing():
            raise ProviderError(
                "COPILOT_PROVIDER_RUNTIME_URL cannot be combined with custom endpoint "
                "routing in runtime.provider; connecting to an existing runtime and "
                "custom endpoint routing are mutually exclusive (the runtime is the endpoint).",
                suggestion=(
                    "Unset COPILOT_PROVIDER_RUNTIME_URL or remove base_url/api_key/"
                    "bearer_token/type/wire_api/headers/azure from runtime.provider."
                ),
                is_retryable=False,
            )
```

File: src/conductor/providers/copilot.py [hunk 4]
```diff
@@ -459,6 +461,41 @@ def _apply_provider_config(self, session_kwargs: dict[str, Any]) -> None:
...
+        settings = self._provider_settings
+
+        url = settings.runtime_url if settings is not None else None
+        url = url or os.environ.get("COPILOT_PROVIDER_RUNTIME_URL")
+
+        token: str | None = None
+        if settings is not None and settings.runtime_token is not None:
+            token = settings.runtime_token.get_secret_value()
+        if token is None:
+            token = os.environ.get("COPILOT_PROVIDER_RUNTIME_TOKEN")
```
Comment (Shazwazza) [inline src/conductor/providers/copilot.py:502]: 🟡 **error_handling**: [R2-002] Runtime connection values resolved from environment variables are not validated like YAML values. An empty `COPILOT_PROVIDER_RUNTIME_URL` (e.g. `${VAR:-}` resolving empty) falls through to a silent nested spawn, and a token-only environment (token set, no URL) silently drops the token and spawns — contradicting the YAML path, which hard-rejects `runtime_token` without `runtime_url`. Both reviewers converged that an empty `COPILOT_PROVIDER_RUNTIME_TOKEN` should NOT be an error: it is the legitimate no-auth / tokenless-runtime case and must normalize to `None`.

**Recommendation**: Normalize/validate env-resolved values before deciding to spawn: strip `runtime_url` and raise a non-retryable `ProviderError` if it is empty-after-strip; raise if a non-empty `runtime_token` is present with no `runtime_url` (mirroring the YAML rule). Normalize an empty `runtime_token` to `None` rather than erroring, preserving the no-auth path. Add provider tests for empty-URL, token-only, and empty-token env combinations.

**Suggestion**:
```python
        if token is None:
            token = os.environ.get("COPILOT_PROVIDER_RUNTIME_TOKEN")

        if url is not None:
            url = url.strip()
            if not url:
                raise ProviderError(
                    "'runtime_url' is empty; remove it or supply a value.",
                    suggestion=(
                        "Set runtime.provider.runtime_url or COPILOT_PROVIDER_RUNTIME_URL "
                        "to a valid port, host:port, or full URL."
                    ),
                    is_retryable=False,
                )
        if token is not None:
            token = token.strip() or None
        if token is not None and url is None:
            raise ProviderError(
                "'runtime_token' requires 'runtime_url' to also be set",
                suggestion=(
                    "Set COPILOT_PROVIDER_RUNTIME_URL alongside "
                    "COPILOT_PROVIDER_RUNTIME_TOKEN, or remove the token."
                ),
                is_retryable=False,
            )

        if url is None:
            return None
        return (url, token)
```

**Comment by Shazwazza** [comment]
# 🤖 Multi-Agent PR Review #2 — Update since Review #1

**PR**: #283 — Add Copilot runtime connection support and update documentation
**Merge readiness**: ⚠️ Needs minor fixes
**Review confidence**: 78/100 🟡 Moderate · *reviewer agreement & finding strength — not a PR-quality score*
**Models**: claude-opus-4.8 (Security & Correctness), gpt-5.5 (Robustness & Completeness)
**Deliberation**: 2 rounds, consensus reached
**Previous review**: Review #1 (referenced for delta comparison)

---

## 📋 Summary

Since Review #1, the author resolved all four prior warnings — the `RuntimeConnection` import was isolated, empty-string `runtime_url` is now rejected at the schema, and the environment-variable activation path enforces the same validation and mutual-exclusion rules as YAML. Two new low-end warnings surfaced as follow-ons to those fixes: the now-isolated import creates a `RuntimeConnection = None` path that `_build_client()` never guards (so an older-but-present SDK hits an opaque `AttributeError` instead of the documented `ProviderError`), and the schema's emptiness checks still let whitespace-only `runtime_url` / `runtime_token` slip through to be silently stripped at runtime. Both are small, well-scoped fixes, so the verdict holds at ⚠️ *Needs minor fixes*.

## 🆕 New Findings

| Severity | File | Finding | Recommendation |
|----------|------|---------|----------------|
| 🟡 | `src/conductor/providers/copilot.py:2132` | `_build_client()` calls `RuntimeConnection.for_uri(...)` unconditionally even though `RuntimeConnection` is now imported as optional (`= None` on `ImportError`). The comment promises "that path raises a clear ProviderError if RuntimeConnection is missing," but no `if RuntimeConnection is None:` guard exists. On an older SDK with a configured `runtime_url` (YAML or `COPILOT_PROVIDER_RUNTIME_URL`), execution raises `AttributeError: 'NoneType' object has no attribute 'for_uri'` instead of the documented, non-retryable `ProviderError`. The existing test only covers the SDK-present path, so it can't catch this. Both reviewers agreed at conviction 9. | Add an explicit `if RuntimeConnection is None:` guard before the `for_uri` call raising a non-retryable `ProviderError` that tells the user to upgrade the Copilot SDK or unset `runtime_url` / `COPILOT_PROVIDER_RUNTIME_URL`. Add a regression test that monkeypatches `RuntimeConnection` to `None`. <details><summary>💡 Suggested fix</summary><br>`if RuntimeConnection is None:` → raise non-retryable `ProviderError` (see diff below).</details> |
| 🟡 | `src/conductor/config/schema.py:1847-1857` | The schema rejects only exact empty strings for secrets and `runtime_url`; whitespace-only YAML values pass validation. The provider later strips them, silently dropping a whitespace-only `runtime_token` to `None` (flipping the auth mode from "use this token" to "no token") or deferring a whitespace-only `runtime_url` failure to runtime — so `conductor validate` misses the bad config despite the schema already promising to reject empty secret/config values. Weak-consensus warning (R1 conceded from suggestion; R2 held warning at conviction 7). | Treat whitespace-only secret values (including `runtime_token`) and `runtime_url` as empty during schema validation using a non-mutating `.strip()` emptiness check that leaves non-empty values untouched, matching the runtime resolver and the schema's own `""` rejection. Add tests for whitespace-only `runtime_url` and `runtime_token`. <details><summary>💡 Suggested fix</summary><br>Non-mutating `.strip() == ""` emptiness check for the four secret fields + `runtime_url` (see diff below).</details> |

<details>
<summary>💡 Suggested fix for the <code>RuntimeConnection is None</code> guard (copilot.py)</summary>

```diff
+        if RuntimeConnection is None:
+            raise ProviderError(
+                "Connecting to an existing Copilot runtime (runtime_url) requires a "
+                "Copilot SDK that provides RuntimeConnection, which is unavailable in "
+                "the installed SDK version.",
+                suggestion=(
+                    "Upgrade the copilot SDK, or unset runtime_url / "
+                    "COPILOT_PROVIDER_RUNTIME_URL to spawn a nested runtime instead."
+                ),
+                is_retryable=False,
+            )
```
</details>

<details>
<summary>💡 Suggested fix for the whitespace-only emptiness check (schema.py)</summary>

```diff
+        for secret_field, value in (
+            ("api_key", self.api_key),
+            ("bearer_token", self.bearer_token),
+            ("auth_token", self.auth_token),
+            ("runtime_token", self.runtime_token),
+        ):
+            if value is not None and value.get_secret_value().strip() == "":
+                raise ValueError(
+                    f"'{secret_field}' is empty; remove the key or supply a value "
+                    "(typo / unset env interpolation?)"
+                )
+        if self.runtime_url is not None and self.runtime_url.strip() == "":
+            raise ValueError(
+                "'runtime_url' is empty; remove the key or supply a value "
+                "(typo / unset env interpolation?)"
+            )
```
</details>

## ✅ Resolved Findings

All four consensus warnings from Review #1 are no longer present.

| Previous Severity | File | Finding | Status |
|-------------------|------|---------|--------|
| 🟡 | `src/conductor/providers/copilot.py:82` | `RuntimeConnection` imported inside the shared SDK-availability `try`, so an older-but-present SDK silently disabled the entire Copilot provider | Resolved |
| 🟡 | `src/conductor/config/schema.py:1881-1904` | Empty-string `runtime_url` not rejected at schema validation (fell back to env and silently spawned a nested runtime) | Resolved |
| 🟡 | `src/conductor/providers/copilot.py:2082` | `COPILOT_PROVIDER_RUNTIME_URL` + custom endpoint routing only warned instead of enforcing mutual exclusion | Resolved |
| 🟡 | `src/conductor/providers/copilot.py:493-497` | Env-resolved runtime values not validated/stripped like YAML (empty URL, token-without-URL) | Resolved |

## 📣 Author Feedback Considered

Findings from Review #1 where the author or a maintainer responded (all four prior threads were marked **resolved**):

| Finding | File:Line | Prior Status | Human Reply (excerpt) | This Review's Action |
|---------|-----------|--------------|------------------------|----------------------|
| R1-001 | `src/conductor/providers/copilot.py` | resolved | — (thread marked resolved, no comment) | Dropped — import isolation landed; no matching finding at this line in Review #2 |
| R1-002 | `src/conductor/config/schema.py:1891` | resolved | — (thread marked resolved, no comment) | Dropped — empty-string `runtime_url` rejection landed |
| R2-001 | `src/conductor/providers/copilot.py:2117` | resolved | — (thread marked resolved, no comment) | Dropped — env-path mutual-exclusion enforcement landed |
| R2-002 | `src/conductor/providers/copilot.py:500` | resolved | — (thread marked resolved, no comment) | Dropped — env-resolved value validation/strip landed |

## 🔄 Changed Findings

No findings changed. The two current findings are new follow-ons rather than re-graded versions of prior findings (different file locations and categories).

## 📊 Delta Summary

- Unchanged findings: 0
- New findings: 2
- Resolved findings: 4
- Changed findings: 0
- Dropped via author feedback: 4
- Downgraded via author feedback: 0
- Re-raised with acknowledgement: 0

## 🏁 Verdict

**Merge readiness**: ⚠️ Needs minor fixes — the four prior warnings are all resolved, but two small follow-on warnings remain: the missing `RuntimeConnection is None` guard in `_build_client()` (R1-001/R2-003) that breaks the documented error contract on older SDKs, and the whitespace-only `runtime_url` / `runtime_token` bypass (R2-004). Both are trivially fixable and have concrete suggested fixes.
**Review confidence**: 78/100 🟡 Moderate · *reviewer agreement & finding strength — not a PR-quality score*

File: src/conductor/providers/copilot.py [hunk 5]
```diff
@@ -2027,6 +2099,38 @@ async def _ensure_client_started(self) -> None:
...
+                    "Unset COPILOT_PROVIDER_RUNTIME_URL or remove base_url/api_key/"
+                    "bearer_token/type/wire_api/headers/azure from runtime.provider."
+                ),
+                is_retryable=False,
+            )
+        logger.info(
+            "Connecting to existing Copilot runtime at %s (no nested runtime spawned)",
+            url,
+        )
+        return CopilotClient(connection=RuntimeConnection.for_uri(url, connection_token=token))
```
Comment (Shazwazza) [inline src/conductor/providers/copilot.py:2135]: 🟡 **error_handling**: [R1-001, R2-003] `_build_client()` dereferences `RuntimeConnection.for_uri(...)` unconditionally even though `RuntimeConnection` is imported as optional for older-but-present Copilot SDKs (RuntimeConnection = None on ImportError). The comment promises 'that path raises a clear ProviderError if RuntimeConnection is missing,' but no `if RuntimeConnection is None:` guard exists. With an older SDK and a configured `runtime_url` (via YAML or `COPILOT_PROVIDER_RUNTIME_URL`), execution raises an opaque `AttributeError: 'NoneType' object has no attribute 'for_uri'` instead of the documented, non-retryable ProviderError. The existing test only exercises the SDK-present path (imports `UriRuntimeConnection`), so it cannot catch this. Both reviewers independently identified this same defect and confirmed it in deliberation at conviction 9.

**Recommendation**: Add an explicit `if RuntimeConnection is None:` guard in `_build_client()` before the `for_uri` call that raises a non-retryable `ProviderError` telling the user to upgrade the Copilot SDK or unset `runtime_url` / `COPILOT_PROVIDER_RUNTIME_URL`, fulfilling the documented contract. Add a regression test that monkeypatches `RuntimeConnection` to `None` and asserts the ProviderError.

```suggestion
        if RuntimeConnection is None:
            raise ProviderError(
                "Connecting to an existing Copilot runtime (runtime_url) requires a "
                "Copilot SDK that provides RuntimeConnection, which is unavailable in "
                "the installed SDK version.",
                suggestion=(
                    "Upgrade the copilot SDK, or unset runtime_url / "
                    "COPILOT_PROVIDER_RUNTIME_URL to spawn a nested runtime instead."
                ),
                is_retryable=False,
            )
```

File: src/conductor/config/schema.py [hunk 6]
```diff
@@ -1806,13 +1842,24 @@ def _check_field_compatibility(self) -> ProviderSettings:
             ("api_key", self.api_key),
             ("bearer_token", self.bearer_token),
             ("auth_token", self.auth_token),
+            ("runtime_token", self.runtime_token),
         ):
             if value is not None and value.get_secret_value() == "":
```
Comment (Shazwazza) [inline src/conductor/config/schema.py]: 🟡 **correctness**: [R2-004] The schema rejects only exact empty strings for secrets and `runtime_url`, but whitespace-only YAML values pass validation. The provider later strips these values, which can silently drop a whitespace-only `runtime_token` to `None` (changing the configured auth mode from 'use this token' to 'no token') or defer a whitespace-only `runtime_url` failure until runtime, so `conductor validate` does not catch the bad configuration despite the schema already promising to reject empty secret/config values. Both reviewers agreed on the fix; they converged on a low-end warning severity (Reviewer 1 conceded from suggestion, accepting a low-end warning; Reviewer 2 held low-end warning at conviction 7).

**Recommendation**: Treat whitespace-only secret values (including `runtime_token`) and `runtime_url` as empty during schema validation, using a non-mutating `.strip()` emptiness check that does not alter non-empty values, so `conductor validate` matches the runtime resolver's behavior and the schema's own '' rejection. Add tests for whitespace-only `runtime_url` and `runtime_token`.

```suggestion
        for secret_field, value in (
            ("api_key", self.api_key),
            ("bearer_token", self.bearer_token),
            ("auth_token", self.auth_token),
            ("runtime_token", self.runtime_token),
        ):
            if value is not None and value.get_secret_value().strip() == "":
                raise ValueError(
                    f"'{secret_field}' is empty; remove the key or supply a value "
                    "(typo / unset env interpolation?)"
                )
        if self.runtime_url is not None and self.runtime_url.strip() == "":
            raise ValueError(
                "'runtime_url' is empty; remove the key or supply a value "
                "(typo / unset env interpolation?)"
            )
```

**Review by Shazwazza — COMMENTED** [review]
🤖 Multi-Agent PR Review #2 — Inline Findings

Posted by the pr-reviewer agent. See the summary comment for the full report.

File: src/conductor/providers/copilot.py (hunk 5 above)
Comment (Shazwazza) [reply src/conductor/providers/copilot.py:2135]: Fixed in 1da010c. Added an explicit if RuntimeConnection is None: guard in _build_client() before the or_uri dereference — it now raises a clear non-retryable ProviderError telling the user to upgrade the SDK or unset runtime_url. Added a regression test that monkeypatches RuntimeConnection to None and asserts the ProviderError.

File: src/conductor/config/schema.py (hunk 6 above)
Comment (Shazwazza) [reply src/conductor/config/schema.py]: Fixed in 1da010c. Whitespace-only secrets (api_key, bearer_token, auth_token, runtime_token) and runtime_url are now treated as empty in schema validation using a non-mutating .strip() check, so conductor validate matches the resolver. Added tests for whitespace-only runtime_url, runtime_token, and api_key.

**Review by jrob5756 — APPROVED** [review]
LGTM. Approved!
